MastodonSDK/Sources/MastodonUI/Scene/ComposeContent/Attachment/AttachmentViewModel+Compress.swift

```diff
@@ -92,3 +93,57 @@ extension AttachmentViewModel {
         }
     }   // end func
 }
+
+extension AttachmentViewModel {
+    @AttachmentViewModelActor
+    func compressImage(data: Data, sizeLimit: SizeLimit) throws -> Output {
+        let maxPayloadSizeInBytes = sizeLimit.image ?? 10 * 1024 * 1024
+
+        guard let image = KFCrossPlatformImage(data: data)?.kf.normalized,
+              var imageData = image.kf.pngRepresentation()
+        else {
+            throw AttachmentError.invalidAttachmentType
+        }
+        
+        repeat {
+            guard let image = KFCrossPlatformImage(data: imageData) else {
+                throw AttachmentError.invalidAttachmentType
+            }
+
+            if imageData.kf.imageFormat == .PNG {
+                // A. png image
+                if imageData.count > maxPayloadSizeInBytes {
+                    guard let compressedJpegData = image.jpegData(compressionQuality: 0.8) else {
+                        throw AttachmentError.invalidAttachmentType
+                    }
+                    os_log("%{public}s[%{public}ld], %{public}s: compress png %.2fMiB -> jpeg %.2fMiB", ((#file as NSString).lastPathComponent), #line, #function, Double(imageData.count) / 1024 / 1024, Double(compressedJpegData.count) / 1024 / 1024)
+                    imageData = compressedJpegData
+                } else {
+                    os_log("%{public}s[%{public}ld], %{public}s: png %.2fMiB", ((#file as NSString).lastPathComponent), #line, #function, Double(imageData.count) / 1024 / 1024)
+                    break
+                }
+            } else {
+                // B. other image
+                if imageData.count > maxPayloadSizeInBytes {
+                    let targetSize = CGSize(width: image.size.width * 0.8, height: image.size.height * 0.8)
+                    let scaledImage = image.kf.resize(to: targetSize)
+                    guard let compressedJpegData = scaledImage.jpegData(compressionQuality: 0.8) else {
+                        throw AttachmentError.invalidAttachmentType
+                    }
+                    os_log("%{public}s[%{public}ld], %{public}s: compress jpeg %.2fMiB -> jpeg %.2fMiB", ((#file as NSString).lastPathComponent), #line, #function, Double(imageData.count) / 1024 / 1024, Double(compressedJpegData.count) / 1024 / 1024)
+                    imageData = compressedJpegData
```

Comment:
I’m worried that encoding a very large image as JPEG repeatedly will result in artifacts. Would it make sense to store the uncompressed image and repeatedly downscale then compress it?

Reply:
```
2022-11-22 20:16:26.117589+0800 Mastodon[32609:6822189] AttachmentViewModel+Compress.swift[119], compressImage(data:sizeLimit:): compress png 192.94MiB -> jpeg 18.46MiB
2022-11-22 20:16:30.005269+0800 Mastodon[32609:6822189] AttachmentViewModel+Compress.swift[133], compressImage(data:sizeLimit:): compress jpeg 18.46MiB -> jpeg 12.87MiB
2022-11-22 20:16:32.362633+0800 Mastodon[32609:6822189] AttachmentViewModel+Compress.swift[133], compressImage(data:sizeLimit:): compress jpeg 12.87MiB -> jpeg 8.81MiB
```

The loop will not be too much. So maybe not a big difference. But If we use the original image in every loop.